File: client/android/BigFoxAndroid/app/src/main/java/vn/com/onesoft/bigfox/io/core/pack/BFPacker.java
package vn.com.onesoft.bigfox.io.core.pack;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Vector;

import vn.com.onesoft.bigfox.io.message.BaseMessage;

/**
 * Created by phamquan on 8/21/15.
 */
public class BFPacker {

    public static int MAX_LENGTH = 100000;

    public static Vector slide(byte[] data) {
        Vector result = new Vector();
        if (data.length <= MAX_LENGTH) {
            result.add(data);
        } else {
            byte[] header = new byte[24];
            System.arraycopy(data, 0, header, 0, 24);
            byte[] body = new byte[data.length - header.length];
            System.arraycopy(data, header.length, body, 0, data.length - header.length);

            for (int i = 0; i < body.length / MAX_LENGTH; i++) {
                byte[] idata = new byte[MAX_LENGTH + header.length];
                System.arraycopy(header, 0, idata, 0, header.length);
                System.arraycopy(body, i * MAX_LENGTH, idata, header.length, MAX_LENGTH);
                normalData(BaseMessage.STATUS_CONTINUE, idata);
                result.add(idata);
            }

            if (data.length > MAX_LENGTH * (data.length / MAX_LENGTH)) {
                int len = body.length - MAX_LENGTH * (body.length / MAX_LENGTH);
                byte[] idata = new byte[len + header.length];
                System.arraycopy(header, 0, idata, 0, header.length);
                System.arraycopy(body, body.length - len, idata, header.length, len);
                normalData(0, idata);
                result.add(idata);
            }

            return result;
        }

        return result;
    }

    private static void normalData(int status, byte[] data) {
        int length = data.length; // 4 byte length, 4 byte tag
        data[0] = (byte) ((length >> 24) & 0x00ff);
        data[1] = (byte) ((length >> 16) & 0x00ff);
        data[2] = (byte) ((length >> 8) & 0x00ff);
        data[3] = (byte) ((length) & 0x00ff);

        data[16] = (byte) ((status >> 24) & 0x00ff);
        data[17] = (byte) ((status >> 16) & 0x00ff);
        data[18] = (byte) ((status >> 8) & 0x00ff);
        data[19] = (byte) ((status) & 0x00ff);
    }
}

File: client/ios/BigFoxClient/BigFoxClient/core/pack/BFPacker.m
//
//  BFPacker.m
//  BigFoxClient
//
//  Created by QuyenNX on 9/25/15.
//  Copyright © 2015 QuyenNX. All rights reserved.
//

#import "BFPacker.h"
#import "BaseMessage.h"

static NSMutableDictionary* mapChannelToByteBuffer = nil ;
@implementation BFPacker

+ (void) initialize {
    if (mapChannelToByteBuffer == nil) {
        mapChannelToByteBuffer = [[NSMutableDictionary alloc]init] ;
    }
}

+ (NSData*) pack:(NSStream*) is : (NSData *)data {
    NSData * dataBuf = [mapChannelToByteBuffer objectForKey:is];
    if (dataBuf == nil) {
        dataBuf = [[NSData alloc] init];
    }
    int length = [data length];
    char* cData = [data bytes];
    int lenResult = dataBuf.length + length;
    char* result = (char*) malloc(dataBuf.length + length);
    if (dataBuf.length > 0) {
        char* buf = [dataBuf bytes];
        for (int i= 0; i< dataBuf.length ; i ++) {
            result[i] = buf[i];
        }
    }
    
    for (int i = dataBuf.length; i < lenResult; i++) {
        result[i] = cData[i - dataBuf.length];
    }
    int status = (int) (((cData[16] & 0xff) << 24) | ((cData[17] & 0xff) << 16) | ((cData[18] & 0xff) << 8) | ((cData[19] & 0xff)));
    if ((status & STATUS_CONTINUE) == 0) {
        result[0] = (char) ((lenResult >> 24) & 0x00ff);
        result[1] = (char) ((lenResult >> 16) & 0x00ff);
        result[2] = (char) ((lenResult >> 8) & 0x00ff);
        result[3] = (char) ((lenResult) & 0x00ff);
        if (lenResult > 10000){
            int a = 0;
        }
        
        [mapChannelToByteBuffer removeObjectForKey:is];
        return [NSData dataWithBytes:result length:lenResult];
    }else{
        [mapChannelToByteBuffer setObject:[NSData dataWithBytes:result length:lenResult] forKey:is];
        return nil;
    }
}

+ (NSArray*) slide : (NSData*) data {
    NSMutableArray* result = [[NSMutableArray alloc] init];
    if ([data length] <= MAX_LENGTH) {
        [result addObject:data];
    }else{
        int length = [data length];
        char* cData = [data bytes];
        char* header = (char*)malloc(24);
        char* body = (char*) malloc(length - 24);
        for (int i= 0; i < length; i++) {
            if (i < 24) {
                header[i] = cData[i];
            }else{
                body[i - 24] = cData[i];
            }
        }
        int lengthBody = length - 24;
        for (int j = 0; j < lengthBody / MAX_LENGTH; j++) {
            int lenData = MAX_LENGTH + 24;
            char* iData = (char*)malloc(lenData);
            for (int k = 0; k < MAX_LENGTH; k ++) {
                if (k < 24) {
                    iData[k] = header[k];
                }else{
                    iData[k] = body[(j * MAX_LENGTH) + k - 24];
                }
            }
            [self normalData:STATUS_CONTINUE :iData : (MAX_LENGTH + 24)];
            [result addObject:[NSData dataWithBytes:iData length:lenData]];
        }
        
        if(length > MAX_LENGTH + (length / MAX_LENGTH)) {
            int len = lengthBody - MAX_LENGTH * (lengthBody / MAX_LENGTH);
            int kLenData = len + 24;
            char* kData = (char*) malloc(kLenData);
            for (int h = 0; h < len; h++) {
                if (h < 24) {
                    kData[h] = header[h];
                }else{
                    kData[h] = body[(lengthBody - len) + h - 24];
                }
            }
            [self normalData:0 :kData :kLenData];
           [result addObject:[NSData dataWithBytes:kData length:kLenData]];
        }
    }
    return result;
}

+ (void) normalData : (int) status : (char*) data : (int)length {
    data[0] = (char) ((length >> 24) & 0x00ff);
    data[1] = (char) ((length >> 16) & 0x00ff);
    data[2] = (char) ((length >> 8) & 0x00ff);
    data[3] = (char) ((length) & 0x00ff);
    
    data[16] = (char) ((status >> 24) & 0x00ff);
    data[17] = (char) ((status >> 16) & 0x00ff);
    data[18] = (char) ((status >> 8) & 0x00ff);
    data[19] = (char) ((status) & 0x00ff);
}

@end
